### PR TITLE
add labels and annotations support for Discovery, TidbMonitor and TidbInitializer

### LINKED ISSUE #4012: Add Pod Annotation Spec for Discovery component
## Feature Request

**Is your feature request related to a problem? Please describe:**

Components such as PD, TiDB, TiKV have already supported set annotations by tc.Spec.Annotations. However, Discovery doesn't support this, we need to add this configuration.
**Describe the feature you'd like:**

**Describe alternatives you've considered:**

**Teachability, Documentation, Adoption, Migration Strategy:**

### PR DESCRIPTION
**Pull request by csuzhangxc**
### What problem does this PR solve?

add labels and annotations support for Discovery, TidbMonitor and TidbInitializer

close #4012

### What is changed and how does it work?
 
copy labels and annotations from CRD to pods.

### Code changes

- [x] Has Go code change
- [ ] Has CI related scripts change

### Tests

- [ ] Unit test 
- [x] E2E test 
- [ ] Manual test 
- [ ] No code 

### Side effects

- [ ] Breaking backward compatibility 
- [ ] Other side effects: 

### Related changes

- [ ] Need to cherry-pick to the release branch 
- [ ] Need to update the documentation 

### Release Notes

Please refer to [Release Notes Language Style Guide](https://github.com/pingcap/tidb-operator/blob/master/docs/release-note-guide.md) before writing the release note.

```release-note
add labels and annotations support for Discovery, TidbMonitor and TidbInitializer
```

### CODE REVIEW

**Comment by csuzhangxc** [comment]
@DanielZhangQD @july2993 PTAL

File: pkg/manager/member/tidb_init_manager.go [hunk 1]
```diff
@@ -463,9 +466,10 @@ func getInitMeta(ti *v1alpha1.TidbInitializer) (metav1.ObjectMeta, label.Label)
 	initLabel := label.NewInitializer().Instance(ti.Name).Initializer(ti.Name)
 
 	objMeta := metav1.ObjectMeta{
-		Name:      name,
-		Namespace: ti.Namespace,
-		Labels:    initLabel,
+		Name:        name,
+		Namespace:   ti.Namespace,
+		Labels:      initLabel,
```
Comment (DanielZhangQD) [inline pkg/manager/member/tidb_init_manager.go:471]: This does not include customized labels.

File: pkg/manager/member/tidb_init_manager.go [hunk 2]
```diff
@@ -347,10 +347,13 @@ func (m *tidbInitManager) makeTiDBInitJob(ti *v1alpha1.TidbInitializer) (*batchv
...
 
 	meta, initLabel := getInitMeta(ti)
+	podLabels := util.CombineStringMap(initLabel, ti.Spec.Labels)
+	podAnnotations := util.CopyStringMap(ti.Spec.Annotations)
 
 	podSpec := &corev1.PodTemplateSpec{
 		ObjectMeta: metav1.ObjectMeta{
-			Labels: initLabel.Labels(),
+			Labels:      podLabels,
+			Annotations: podAnnotations,
```
Comment (DanielZhangQD) [inline pkg/manager/member/tidb_init_manager.go:356]: Is it OK to use `meta.Annotations` here?

File: pkg/controller/generic_control.go [hunk 3]
```diff
@@ -166,6 +166,9 @@ func (w *typedWrapper) CreateOrUpdateDeployment(controller runtime.Object, deplo
 			}
 		}
 		// pod selector of deployment is immutable, so we don't mutate the labels of pod
```
[the remark this answers is not in the document]
Comment (DanielZhangQD) [reply pkg/controller/generic_control.go:168]: Please also test the case that the labels/annotations can be updated.

File: pkg/monitor/monitor/util.go [hunk 4]
```diff
@@ -1045,12 +1049,12 @@ func getMonitorStatefulSetSkeleton(sa *core.ServiceAccount, monitor *v1alpha1.Ti
...
 			},
 			Selector: &meta.LabelSelector{
-				MatchLabels: buildTidbMonitorLabel(monitor.Name),
+				MatchLabels: stsLabels,
 			},
 			Template: core.PodTemplateSpec{
 				ObjectMeta: meta.ObjectMeta{
-					Labels:      buildTidbMonitorLabel(monitor.Name),
-					Annotations: util.CopyStringMap(monitor.Spec.Annotations),
+					Labels:      podLabels,
```
Comment (DanielZhangQD) [inline pkg/monitor/monitor/util.go:1056]: Please also support the labels and annotations for the services created for Tidbmonitor。

File: pkg/apis/pingcap/v1alpha1/tidbinitializer_types.go [hunk 5]
```diff
@@ -84,6 +84,12 @@ type TidbInitializerSpec struct {
 	// +optional
 	PasswordSecret *string `json:"passwordSecret,omitempty"`
 
+	// +optional
```
[the remark this answers is not in the document]
Comment (csuzhangxc) [reply pkg/apis/pingcap/v1alpha1/tidbinitializer_types.go]: https://github.com/pingcap/tidb-operator/pull/4029/commits/3ea7b432dc9678acd1f63de8d00b47553b4ddf7a

File: pkg/manager/member/tidb_init_manager.go (hunk 1 above)
Comment (csuzhangxc) [reply pkg/manager/member/tidb_init_manager.go]: https://github.com/pingcap/tidb-operator/pull/4029/commits/3ea7b432dc9678acd1f63de8d00b47553b4ddf7a

File: pkg/manager/member/tidb_init_manager.go (hunk 2 above)
Comment (csuzhangxc) [reply pkg/manager/member/tidb_init_manager.go]: YES, https://github.com/pingcap/tidb-operator/pull/4029/commits/3ea7b432dc9678acd1f63de8d00b47553b4ddf7a

File: pkg/monitor/monitor/util.go (hunk 4 above)
Comment (csuzhangxc) [reply pkg/monitor/monitor/util.go:1056]: https://github.com/pingcap/tidb-operator/pull/4029/commits/3ea7b432dc9678acd1f63de8d00b47553b4ddf7a

File: pkg/controller/generic_control.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (csuzhangxc) [reply pkg/controller/generic_control.go:168]: if update labels here, I got the following error:

```
I0624 09:56:25.479207       1 tidb_cluster_controller.go:124] TidbCluster: tidb-cluster-8488/basic-v3-0-20, still need sync: error creating or updating discovery service: Deployment.apps "basic-v3-0-20-discovery" is invalid: spec.template.metadata.labels: Invalid value: map[string]string{"app.kubernetes.io/component":"discovery", "app.kubernetes.io/instance":"basic-v3-0-20", "app.kubernetes.io/managed-by":"tidb-operator", "app.kubernetes.io/name":"tidb-cluster", "cluster-test-key":"new-value"}: `selector` does not match template `labels`, requeuing
```

but the label selector is immutable, ref https://kubernetes.io/docs/concepts/workloads/controllers/deployment/#label-selector-updates

**Comment by csuzhangxc** [comment]
@DanielZhangQD PTAL again

File: pkg/controller/generic_control.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (DanielZhangQD) [reply pkg/controller/generic_control.go:168]: OK, get it.

File: pkg/manager/member/tidb_discovery_manager.go [hunk 6]
```diff
@@ -249,6 +249,8 @@ func (m *realTidbDiscoveryManager) getTidbDiscoveryDeployment(obj metav1.Object)
 		})
 	}
 
+	podLabels := util.CombineStringMap(l.Labels(), baseSpec.Labels())
```
Comment (DanielZhangQD) [inline pkg/manager/member/tidb_discovery_manager.go:252]: `getDiscoveryMeta` function also needs to update?

File: pkg/monitor/monitor/util.go [hunk 7]
```diff
@@ -1030,13 +1030,17 @@ func getMonitorStatefulSetSkeleton(sa *core.ServiceAccount, monitor *v1alpha1.Ti
 		replicas = *monitor.Spec.Replicas
 	}
 	name := GetMonitorObjectName(monitor)
+	stsLabels := buildTidbMonitorLabel(monitor.Name)
+	podLabels := util.CombineStringMap(stsLabels, monitor.Spec.Labels)
```
[the remark this answers is not in the document]
Comment (csuzhangxc) [reply pkg/monitor/monitor/util.go:1034]: changed to return a newly created map in https://github.com/pingcap/tidb-operator/pull/4029/commits/b6b36c266ce358ced2c4fb1b7b4ddc355545efc1

File: pkg/util/object_meta.go [hunk 8]
```diff
@@ -16,19 +16,14 @@ package util
 // Utilities here are used for modifying the ObjectMeta field for k8s object
 
 // CombineStringMap merges maps to the left
-func CombineStringMap(a map[string]string, maps ...map[string]string) map[string]string {
```
Comment (DanielZhangQD) [inline pkg/util/object_meta.go:19]: Should we make sure that the values in the map `a` should not be overridden by the values in the `maps`?

Comment (csuzhangxc) [reply pkg/util/object_meta.go:19]: changed to keep the first value in https://github.com/pingcap/tidb-operator/pull/4029/commits/437715ec914f7e74a236835467309e9470dd2d0b

File: pkg/manager/member/tidb_init_manager.go [hunk 9]
```diff
@@ -350,7 +350,8 @@ func (m *tidbInitManager) makeTiDBInitJob(ti *v1alpha1.TidbInitializer) (*batchv
 
 	podSpec := &corev1.PodTemplateSpec{
 		ObjectMeta: metav1.ObjectMeta{
-			Labels: initLabel.Labels(),
+			Labels:      initLabel.Labels(),
```
[the remark this answers is not in the document]
Comment (csuzhangxc) [reply pkg/manager/member/tidb_init_manager.go]: fixed in https://github.com/pingcap/tidb-operator/pull/4029/commits/f205edafe908f7b5944a47e06ce7807b38928681.

File: pkg/monitor/monitor/util.go [hunk 10]
```diff
@@ -870,9 +870,9 @@ func getMonitorService(monitor *v1alpha1.TidbMonitor) []*core.Service {
 			ObjectMeta: meta.ObjectMeta{
 				Name:            grafanaName(monitor),
 				Namespace:       monitor.Namespace,
-				Labels:          grafanaLabel.Labels(),
+				Labels:          util.CombineStringMap(grafanaLabel.Labels(), monitor.Spec.Labels, monitor.Spec.Grafana.Service.Labels),
 				OwnerReferences: []meta.OwnerReference{controller.GetTiDBMonitorOwnerRef(monitor)},
-				Annotations:     monitor.Spec.Grafana.Service.Annotations,
+				Annotations:     util.CombineStringMap(monitor.Spec.Grafana.Service.Annotations, monitor.Spec.Annotations),
```
[the remark this answers is not in the document]
Comment (csuzhangxc) [reply pkg/monitor/monitor/util.go:875]: https://github.com/pingcap/tidb-operator/commit/b6b36c266ce358ced2c4fb1b7b4ddc355545efc1

File: pkg/monitor/monitor/util.go [hunk 11]
```diff
@@ -793,9 +793,9 @@ func getMonitorService(monitor *v1alpha1.TidbMonitor) []*core.Service {
 		ObjectMeta: meta.ObjectMeta{
 			Name:            prometheusName,
 			Namespace:       monitor.Namespace,
-			Labels:          promeLabel.Labels(),
+			Labels:          util.CombineStringMap(promeLabel.Labels(), monitor.Spec.Labels, monitor.Spec.Prometheus.Service.Labels),
 			OwnerReferences: []meta.OwnerReference{controller.GetTiDBMonitorOwnerRef(monitor)},
-			Annotations:     monitor.Spec.Prometheus.Service.Annotations,
+			Annotations:     util.CombineStringMap(monitor.Spec.Prometheus.Service.Annotations, monitor.Spec.Annotations),
```
[the remark this answers is not in the document]
Comment (csuzhangxc) [reply pkg/monitor/monitor/util.go:798]: https://github.com/pingcap/tidb-operator/commit/b6b36c266ce358ced2c4fb1b7b4ddc355545efc1

File: pkg/manager/member/tidb_discovery_manager.go (hunk 6 above)
Comment (csuzhangxc) [reply pkg/manager/member/tidb_discovery_manager.go:252]: our other statefulset didn't include user-specified labels and annotations yet. so let this deployment as is now, if needed, we can added later.

**Comment by csuzhangxc** [comment]
/test pull-e2e-kind-serial

for restarting Docker failed

File: pkg/manager/member/tidb_discovery_manager.go (hunk 6 above)
Comment (DanielZhangQD) [reply pkg/manager/member/tidb_discovery_manager.go:252]: ok

File: pkg/monitor/monitor/util.go [hunk 12]
```diff
@@ -793,9 +793,9 @@ func getMonitorService(monitor *v1alpha1.TidbMonitor) []*core.Service {
 		ObjectMeta: meta.ObjectMeta{
 			Name:            prometheusName,
 			Namespace:       monitor.Namespace,
-			Labels:          promeLabel.Labels(),
+			Labels:          util.CombineStringMap(promeLabel.Labels(), monitor.Spec.Labels, monitor.Spec.Prometheus.Service.Labels),
```
Comment (DanielZhangQD) [inline pkg/monitor/monitor/util.go:796]:
```suggestion
			Labels:          util.CombineStringMap(promeLabel.Labels(), monitor.Spec.Prometheus.Service.Labels, monitor.Spec.Labels),
```

File: pkg/monitor/monitor/util.go [hunk 13]
```diff
@@ -837,9 +837,9 @@ func getMonitorService(monitor *v1alpha1.TidbMonitor) []*core.Service {
 		ObjectMeta: meta.ObjectMeta{
 			Name:            reloaderName,
 			Namespace:       monitor.Namespace,
-			Labels:          buildTidbMonitorLabel(monitor.Name),
+			Labels:          util.CombineStringMap(buildTidbMonitorLabel(monitor.Name), monitor.Spec.Labels, monitor.Spec.Reloader.Service.Labels),
```
Comment (DanielZhangQD) [inline pkg/monitor/monitor/util.go:840]: ditto

File: pkg/controller/generic_control.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (csuzhangxc) [reply pkg/controller/generic_control.go]: https://github.com/pingcap/tidb-operator/pull/4029/commits/8f70f9e1844720d76a6d6743512d4fb73431ea76

**Review by DanielZhangQD — APPROVED** [review]
LGTM

**Review by july2993 — APPROVED** [review]

**Pull request merged**